finesse/hardware/temperature/dp9800.py

```diff
@@ -0,0 +1,219 @@
+            print(f"Instrument type: {instr_type}")
+            print(f"Logging: {logging_state}")
+            print(f"Autoscan: {scanning_state}")
+            print(f"Audible button: {audible_state}")
+            print(f"Temperature unit: {temp_unit}")
+
+    def read(self) -> bytes:
+        """Read temperature data from the DP9800.
+
+        The DP9800 returns a sequence of bytes containing the
+        temperatures measured on each channel, in the format
+
+          STX T SP t1 SP t2 SP t3 SP t4 SP t5 SP t6 SP t7 SP t8 SP t9 ff ETX BCC NUL
+
+        where
+            t1, t2, ..., t9: temperature values in the format %4.2f
+            STX: Start of Text (ASCII 2)
+            ETX: End of Text (ASCII 3)
+            NUL: Null character (ASCII 0)
+            SP: Space (ASCII 32)
+            BCC: Block Check Character
+            ff: System flag in hexadecimal
+
+        For error checking, the BCC is calculated by performing consecutive XOR
+        operations on the message and compared with the BCC received.
+
+        Note a slight peculiarity:
+        The device actually returns 9 temperatures, while the documentation
+        states that there should be 8. The device shows 8 values, corresponding
+        to the values with indices 1 to 8. The first value (at index 0) is
+        therefore ignored.
+
+        Returns:
+            data: the sequence of bytes read from the device
+
+        Raises:
+            DP9800Error: Malformed message received from device
+        """
+        if self.serial.in_waiting == 0:
+            raise DP9800Error("No data waiting to be read")
+
+        try:
+            data = self.serial.read_until(b"\x00")
+        except SerialTimeoutException:
+            raise DP9800Error("Read request timed out")
+
+        return data
+
+    def request_read(self) -> None:
+        """Write a message to the DP9800 to prepare for a read operation.
+
+        Format:
+
+          EOT T ENQ
+
+        EOT: End of Transmission (ASCII 4)
+        ENQ: Enquiry (ASCII 5)
+
+        Raises:
+            DP9800Error: Error writing to the device
+        """
+        try:
+            self.serial.write(b"\x04T\x05")
+        except Exception as e:
+            raise DP9800Error(e)
+
+    def send_temperatures(self) -> None:
+        """Perform the complete process of reading from the DP9800.
+
+        Writes to the DP9800 requesting a read operation.
+        Reads the raw data from the DP9800.
+        Parses the data and broadcasts the temperatures.
+        """
+        time_now = datetime.now().timestamp()
+        try:
+            self.request_read()
+            data = self.read()
+            check_data(data)
+            temperatures, _ = parse_data(data)
+            pub.sendMessage(
+                "temperature_monitor.data.response", values=temperatures, time=time_now
+            )
+        except Exception as e:
+            self._error_occurred(e)
```

Comment:
As above, I'd be wary about wrapping code in a `try`/`except` block when it doesn't need to be. The danger here is that if an exception occurs in one of the pubsub listeners, it will percolate up the stack and `self._error_occurred` will be called, which will make debugging annoying.

```suggestion
        except Exception as e:
            self._error_occurred(e)
        else:
            pub.sendMessage(
                "temperature_monitor.data.response", values=temperatures, time=time_now
            )
```